File: 2024 1월 3주차/백준 1707.py
from collections import deque

def bfs(graph, start, visited):
    queue = deque([start])
    visited[start] = 1 
    
    while queue:
        node = queue.popleft()
        for i in graph[node]:
            if visited[i] == 0:  
                visited[i] = -visited[node]  
                queue.append(i)
            elif visited[i] == visited[node]:  
                return False  
    return True

K = int(input())

for _ in range(K):
    V, E = map(int, input().split())
    graph = [[] for _ in range(V + 1)]
    visited = [0] * (V + 1)
    
    for _ in range(E):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)

    is_bipartite = True
    for i in range(1, V + 1):
        if visited[i] == 0:
            if not bfs(graph, i, visited): 
                is_bipartite = False
                break

    print("YES" if is_bipartite else "NO")

File: 2024 1월 3주차/백준 1707.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let input = [];
let lineNumber = 0;

function bfs(graph, start, visited) {
    let queue = [];
    queue.push(start);
    visited[start] = 1;

    while (queue.length > 0) {
        let node = queue.shift();
        for (let adjNode of graph[node]) {
            if (visited[adjNode] === 0) {
                visited[adjNode] = -visited[node];
                queue.push(adjNode);
            } else if (visited[adjNode] === visited[node]) {
                return false;
            }
        }
    }
    return true;
}

rl.on('line', function (line) {
    input.push(line);
    lineNumber++;
}).on('close', function () {
    let currentLine = 0;
    let testCases = parseInt(input[currentLine++]);
    for (let t = 0; t < testCases; t++) {
        let [V, E] = input[currentLine++].split(' ').map(x => parseInt(x));
        let graph = Array.from({ length: V + 1 }, () => []);
        let visited = new Array(V + 1).fill(0);

        for (let i = 0; i < E; i++) {
            let [u, v] = input[currentLine++].split(' ').map(x => parseInt(x));
            graph[u].push(v);
            graph[v].push(u);
        }

        let isBipartite = true;
        for (let i = 1; i <= V; i++) {
            if (visited[i] === 0 && !bfs(graph, i, visited)) {
                isBipartite = false;
                break;
            }
        }
        console.log(isBipartite ? 'YES' : 'NO');
    }
    process.exit();
});